src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Imaging/ImageMonikerDebuggerDisplay.cs

```diff
@@ -0,0 +1,50 @@
+﻿// Licensed to the .NET Foundation under one or more agreements. The .NET Foundation licenses this file to you under the MIT license. See the LICENSE.md file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using Microsoft.VisualStudio.Imaging;
+using Microsoft.VisualStudio.ProjectSystem;
+
+using ImageMoniker = Microsoft.VisualStudio.Imaging.Interop.ImageMoniker;
+
+[assembly: DebuggerDisplay("{Microsoft.VisualStudio.ProjectSystem.Imaging.ImageMonikerDebuggerDisplay.FromImageMoniker(this)}", Target = typeof(ImageMoniker))]
+[assembly: DebuggerDisplay("{Microsoft.VisualStudio.ProjectSystem.Imaging.ImageMonikerDebuggerDisplay.FromProjectImageMoniker(this)}", Target = typeof(ProjectImageMoniker))]
+
+namespace Microsoft.VisualStudio.ProjectSystem.Imaging
+{
+    /// <summary>
+    ///     Provides a friendly display for <see cref="ProjectImageMoniker"/> and <see cref="ImageMoniker"/> instances
+    ///     based on the well known values from <see cref="KnownImageIds"/>.
+    /// </summary>
+    internal static class ImageMonikerDebuggerDisplay
+    {
+        private static readonly Dictionary<(Guid, int), string> s_displayNames = CalculateDisplayNames();
+
+        internal static string FromImageMoniker(ImageMoniker moniker) => DebugDisplay(moniker.Guid, moniker.Id);
+
+        internal static string FromProjectImageMoniker(ProjectImageMoniker moniker) => DebugDisplay(moniker.Guid, moniker.Id);
+
+        private static string DebugDisplay(Guid guid, int id)
+        {
+            if (s_displayNames.TryGetValue((guid, id), out string displayName))
+            {
+                return displayName;
+            }
+
+            return $"{guid} ({id})";
+        }
+
+        private static Dictionary<(Guid, int), string> CalculateDisplayNames()
+        {
+            Type type = typeof(KnownImageIds);
+
+            return type.GetFields()
+                       .Where(field => field.IsLiteral && field.FieldType == typeof(int))
+                       .ToDictionary(field =>
+                            (KnownImageIds.ImageCatalogGuid, (int)field.GetRawConstantValue()),
+                            field => $"{type.Name}.{field.Name}");
```

Comment:
Does this only run when debugging? We don't want the startup and allocation overhead unless it's used. I'm unclear when a type used with `DebuggerDisplayAttribute` is initialized.

Reply:
Spoke with @davkean offline about this and the type initializer won't run until needed by the debugger, so there's no risk of initializing this type and running this code in non-debugger situations.